### PR TITLE
hwr: add per-level fog settings

### LINKED ISSUE #118: Feature Request: Per-Level Distance Fog settings
This is TRLE related.
It would be nice to have a per-level setting of the Distance Fog and Cull settings.
TombATI had a setting for that but only for the whole game.

### PR DESCRIPTION
**Pull request by rr-**
Closes #118.

### CODE REVIEW

**Comment by oziphantom** [comment]
Should the water stuff be part of this commit?

**Comment by rr-** [comment]
No, I don't think so, I only ported GLRage's distance settings override patch. Plus #145 probably will require different approach.

**Review by thatguyfrombb — COMMENTED** [review]
Default settings in this PR makes the wall/sprite popping very apparent. This is also an issue on TombATI when changing the draw distance. Otherwise, the values in the JSON can be negative, which makes for some really cool effects. If there's no foreseeable issue with that, I'd love to leave it in.
Causes popping issues

**Comment by rr-** [comment]
What would be the default values you recommend @thatguyfrombb ?

**Comment by thatguyfrombb** [comment]
Fade at `16384` and max at `24576` is the ceiling for absolutely safe settings. Anything over that and the sprites popping in are visible.
Fade at `20480` and max at `28672` is the limit for buildings/walls. Sprites already pop at visible distance, but walls fade in nicely. Anything over this and walls start popping in.

With the current code, I'd go for `16384` so it should just work everywhere and add a note in the README so users can change it at will. Whenever we get it to render better we can change our defaults.

**Comment by rr-** [comment]
Sounds good. I'll amend the settings. Could you create two new issues regarding the hard z limit for sprites and for the geometry?

**Comment by rr-** [comment]
Done, but haven't tested the new settings

**Comment by thatguyfrombb** [comment]
So I talked about TombATI from what I remembered it doing. Turns out, TombATI does handle the original values it ships with correctly. I remembered it having issues because I tried very large values. So, this is *our* bug. Will investigate tomorrow.

**Comment by rr-** [comment]
@thatguyfrombb should be good now, improved on multiple ends
- distances in the config are now in tiles, not in pixels
- ported tr2main's CalculateFogShade to support fog depths other than 8 tiles
- fixed the clipping issues

File: cfg/Tomb1Main_gameflow.json5 [hunk 1]
```diff
@@ -19,6 +19,17 @@
     // [0.45, 1.0, 1.0] is default TombATI filter.
     "water_color": [0.45, 1.0, 1.0],
 
+    // The distance (in tiles) at which objects and the world
+    // start to fade into blackness.
+    // The default hardcoded value in TR1 is 12.
+    // The default hardcoded value in TombATI is 80.
```
Comment (thatguyfrombb) [inline cfg/Tomb1Main_gameflow.json5:25]: In TombATI fade is 72, not 80, and the whole feature is off by default to "not ruin the ambience" or something like that.

File: src/game/gameflow.c [hunk 2]
```diff
@@ -201,6 +201,30 @@ static int8_t S_LoadScriptMeta(struct json_object_s *obj)
             json_array_get_number_double(tmp_arr, 2, GF.water_color.b);
     }
 
+    if (json_object_get_value(obj, "draw_distance_fade")) {
+        double value = json_object_get_number_double(
+            obj, "draw_distance_fade", JSON_INVALID_NUMBER);
+        if (value == JSON_INVALID_NUMBER) {
+            LOG_ERROR("'draw_distance_fade' must be an integer");
```
Comment (thatguyfrombb) [inline src/game/gameflow.c:208]: Incorrect error message, we take doubles now

File: src/game/gameflow.c [hunk 3]
```diff
@@ -201,6 +201,30 @@ static int8_t S_LoadScriptMeta(struct json_object_s *obj)
...
+        GF.draw_distance_fade = value;
+    } else {
+        GF.draw_distance_fade = 12;
+    }
+
+    if (json_object_get_value(obj, "draw_distance_max")) {
+        double value = json_object_get_number_double(
+            obj, "draw_distance_max", JSON_INVALID_NUMBER);
+        if (value == JSON_INVALID_NUMBER) {
+            LOG_ERROR("'draw_distance_max' must be an integer");
```
Comment (thatguyfrombb) [inline src/game/gameflow.c:220]: Same as above

File: src/game/gameflow.c [hunk 4]
```diff
@@ -201,6 +201,30 @@ static int8_t S_LoadScriptMeta(struct json_object_s *obj)
...
+            return 0;
+        }
+        GF.draw_distance_max = value;
+    } else {
+        GF.draw_distance_max = 20;
```
Comment (thatguyfrombb) [inline src/game/gameflow.c:225]: It's stored in a float, let's make that clear by assigning `20.0f`

File: cfg/Tomb1Main_gameflow.json5 [hunk 5]
```diff
@@ -19,6 +19,17 @@
...
+    "draw_distance_fade": 22,
```
Comment (thatguyfrombb) [inline cfg/Tomb1Main_gameflow.json5:26]: Should we change it to `22.0` and max to `30.0` to show that it accepts floats?

**Review by thatguyfrombb — CHANGES_REQUESTED** [review]
Testing works great!

File: src/game/gameflow.c (hunk 2 above)
Comment (rr-) [reply src/game/gameflow.c]: Fixed

File: cfg/Tomb1Main_gameflow.json5 (hunk 1 above)
Comment (rr-) [reply cfg/Tomb1Main_gameflow.json5]: Fixed

File: src/game/gameflow.c (hunk 3 above)
Comment (rr-) [reply src/game/gameflow.c]: Fixed

File: src/game/gameflow.c (hunk 4 above)
Comment (rr-) [reply src/game/gameflow.c]: Fixed

File: cfg/Tomb1Main_gameflow.json5 (hunk 5 above)
Comment (rr-) [reply cfg/Tomb1Main_gameflow.json5]: Good call, changed

**Comment by thatguyfrombb** [comment]
Nearly forgot, do we want this in README?

**Comment by rr-** [comment]
Yes, because the per-level functionality wasn't there originally. I'll add it

**Pull request merged**